caas/kubernetes/provider/customresourcedefinitions.go

```diff
@@ -82,3 +107,224 @@ func (k *kubernetesClient) deleteCustomResourceDefinitions(appName string) error
+	crds, err := k.extendedCient().ApiextensionsV1beta1().CustomResourceDefinitions().List(v1.ListOptions{IncludeUninitialized: true})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, crd := range crds.Items {
+		for _, version := range crd.Spec.Versions {
+			crdClient, err := k.getCustomResourceDefinitionClient(&crd, version.Name)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = crdClient.DeleteCollection(&v1.DeleteOptions{
+				PropagationPolicy: &defaultPropagationPolicy,
+			}, v1.ListOptions{
+				LabelSelector:        labelsToSelector(k.getCRLabels(appName)),
+				IncludeUninitialized: true,
+			})
+			if err != nil && !k8serrors.IsNotFound(err) {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+type apiVersionGetter interface {
+	GetAPIVersion() string
+}
+
+func getCRVersion(cr apiVersionGetter) string {
+	ss := strings.Split(cr.GetAPIVersion(), "/")
+	return ss[len(ss)-1]
+}
+
+func (k *kubernetesClient) ensureCustomResources(
+	appName string,
+	crSpecs map[string][]unstructured.Unstructured,
+) (cleanUps []func(), _ error) {
+	crds, err := k.getCRDsForCRs(crSpecs)
+	if err != nil {
+		return cleanUps, errors.Trace(err)
+	}
+
+	for crdName, crSpecList := range crSpecs {
+		crd, ok := crds[crdName]
+		if !ok {
+			// This should not happen.
+			return cleanUps, errors.NotFoundf("custom resource definition %q", crdName)
+		}
+		for _, crSpec := range crSpecList {
+			crdClient, err := k.getCustomResourceDefinitionClient(crd, getCRVersion(&crSpec))
+			if err != nil {
+				return cleanUps, errors.Trace(err)
+			}
+			crSpec.SetLabels(k.getCRLabels(appName))
+			_, crCleanUps, err := ensureCustomResource(crdClient, &crSpec)
+			cleanUps = append(cleanUps, crCleanUps...)
+			if err != nil {
+				return cleanUps, errors.Annotate(err, fmt.Sprintf("ensuring custom resource %q", crSpec.GetName()))
+			}
+			logger.Debugf("ensured custom resource %q", crSpec.GetName())
+		}
+	}
+	return cleanUps, nil
+}
+
+func ensureCustomResource(api dynamic.ResourceInterface, cr *unstructured.Unstructured) (out *unstructured.Unstructured, cleanUps []func(), err error) {
+	logger.Debugf("creating custom resource %q", cr.GetName())
+	if out, err = api.Create(cr); err == nil {
+		cleanUps = append(cleanUps, func() {
+			deleteCustomResourceDefinition(api, out.GetName(), out.GetUID())
+		})
+		return out, cleanUps, nil
+	}
+	if !k8serrors.IsAlreadyExists(err) {
+		return nil, cleanUps, errors.Trace(err)
+	}
+	// K8s complains about metadata.resourceVersion is required for an update, so get it before updating.
+	existingCR, err := api.Get(cr.GetName(), v1.GetOptions{})
+	if err != nil {
+		return nil, cleanUps, errors.Trace(err)
+	}
+	cr.SetResourceVersion(existingCR.GetResourceVersion())
+	logger.Debugf("updating custom resource %q", cr.GetName())
+	// TODO(caas): do label check to ensure the resource to be updated was created by Juju once caas upgrade steps of 2.7 in place.
```

Comment:
Don't need this todo because we didn't have custom resources until now

Reply:
removed this TODO, because we are using resource version when doing `update`.
So it's safe.